Improve chat functionality of mediation/arbitration

jmacxx opened the pull request:
- Show badge with number of unread chat messages on each ticket
- Use icons for chat, info and process ticket functions
- Select a row by clicking on it, no clunky select button needed.
- Show chat messages in a movable/sizable popup window.
- More space for ticket list.
- [x] @leo816 request : indicate if the trade period is over
- [x] @leo816 request : indicate the XMR tx proof
- [x] Indicate more clearly the sender of each chat message: https://github.com/bisq-network/bisq/issues/3412
- [x] @ripcurlx request for Support badge count: https://github.com/bisq-network/bisq/pull/5207#issuecomment-781967447
- [x] @pazza83 Allow trader to close own mediation ticket if the trade is paid out (because in odd cases they do not get closed on the trader's side).

Fixes #5097 
Fixes #3412


---


Agent view:

![image](https://user-images.githubusercontent.com/47253594/109433296-49ea0300-79d5-11eb-90ec-d40d6dca0974.png)

Client view:

![image](https://user-images.githubusercontent.com/47253594/109432992-8c124500-79d3-11eb-8d73-a3cade4c970c.png)


Light mode:

![image](https://user-images.githubusercontent.com/47253594/109433414-d5639400-79d5-11eb-8c10-51cf4bc9c37f.png)

---

@leo816
Indicate trade end period
Indicate XMR tx proof (when applicable)

![image](https://user-images.githubusercontent.com/47253594/109433353-8453a000-79d5-11eb-93a6-f58ae14959e5.png)

---

@pazza83
Allow trader to close ticket if paid out

![image](https://user-images.githubusercontent.com/47253594/109433464-080d8c80-79d6-11eb-8fbe-04966a7f1d7d.png)
![image](https://user-images.githubusercontent.com/47253594/109433475-1491e500-79d6-11eb-889e-fd72cd6aa22c.png)

---

Indicate more clearly the sender (trader side):

![image](https://user-images.githubusercontent.com/47253594/109433575-b0235580-79d6-11eb-817c-5532e6d1f732.png)


Indicate more clearly the sender (agent side):

![image](https://user-images.githubusercontent.com/47253594/109433557-941fb400-79d6-11eb-9654-465f8422f1aa.png)

Conza88 commented:
Too late to incorporate https://github.com/bisq-network/bisq/issues/5044 ? hah

ripcurlx commented:
@jmacxx Great that you picked this up!

IMO we should only have the info icon in front and move the other action icons (process and chat) to the end. It just feels weird for me having this action icons in the front (yes, I know we had the CTA buttons there as well in the past). What do you think?

Also it would be great if you could adapt the badge number on the support button so it doesn't reflect the number of open support cases, but rather if there is an action that needs to be done. e.g. if there are new support cases add a `new` badge to the info icon (add the number of new support cases to the support button) and if it was selected once decrease the badge number and remove the `new` badge. If there are new chat message for a support case add (+1) to the support button badge and decrease it (-1) if the chat messages were viewed. If a support case was closed add +1 and decrease it if the final message by the mediator/arbitrator was viewed and so on. Do you know what I mean?

I think doing like that prevents that people are ignoring the badge icon as it is always there no matter if there is an action to do or not.

Commit:
Improve chat functionality of mediation/arbitration

Show badge with number of unread chat messages on each ticket
Use icons for chat, info and process ticket functions
Select a row by clicking on it, no clunky select button needed.
Show chat messages in a movable popup window.
More space for ticket list.

Implement requested feature additions:

Indicate if the trade period is over
Indicate more clearly the sender of each chat message
Support badge count
Indicate the XMR tx proof (when applicable)
Allow trader to close own mediation ticket if the trade is paid out

Fixes:

null check for cases when extraData null/not applicable
when upgrading closed disputes, clear chat unread count

ghost commented:
Ready for review!

Dispute object now contains a `State` field to represent whether it is new, open, reopened, or closed => necessary for the count on the Support badge and displaying a "New" badge on table rows.

ripcurlx commented:
@jmacxx Is there a reason why you chose orange for the new badge?

Commit:
color of new badge should be red

ghost commented:
> @jmacxx Is there a reason why you chose orange for the new badge?

You're right, it should be red.  Fix pushed.

![image](https://user-images.githubusercontent.com/47253594/110830162-85e75880-825e-11eb-89a0-27ea3b5b150f.png)

Conza88 commented:
Awesome work

ripcurlx commented on desktop/src/main/java/bisq/desktop/main/overlays/windows/DisputeSummaryWindow.java:
```diff
@@ -104,6 +104,7 @@
     private final TradeWalletService tradeWalletService;
     private final BtcWalletService btcWalletService;
     private final TxFeeEstimationService txFeeEstimationService;
+    // PR5160 will add private final MempoolService mempoolService;
```
Why do you want to add this comment?

ripcurlx commented on desktop/src/main/java/bisq/desktop/main/overlays/windows/DisputeSummaryWindow.java:
```diff
@@ -141,6 +143,7 @@ public DisputeSummaryWindow(@Named(FormattingUtils.BTC_FORMATTER_KEY) CoinFormat
                                 TradeWalletService tradeWalletService,
                                 BtcWalletService btcWalletService,
                                 TxFeeEstimationService txFeeEstimationService,
+                                // PR5160 will add MempoolService mempoolService,
```
Same here: why add this comment?

ripcurlx commented on desktop/src/main/java/bisq/desktop/main/overlays/windows/DisputeSummaryWindow.java:
```diff
@@ -149,6 +152,7 @@ public DisputeSummaryWindow(@Named(FormattingUtils.BTC_FORMATTER_KEY) CoinFormat
         this.tradeWalletService = tradeWalletService;
         this.btcWalletService = btcWalletService;
         this.txFeeEstimationService = txFeeEstimationService;
+        // PR5160 will add this.mempoolService = mempoolService;
```
and same here

ripcurlx commented on desktop/src/main/java/bisq/desktop/main/overlays/windows/DisputeSummaryWindow.java:
```diff
@@ -314,6 +303,26 @@ private void addInfoPane() {
+        if (isMediationDispute) {
+            if (dispute.getTradePeriodEnd().getTime() > 0) {
+                String status = DisplayUtils.formatDateTime(dispute.getTradePeriodEnd());
+                Label tpe = addConfirmationLabelLabel(gridPane, ++rowIndex, Res.get("disputeSummaryWindow.tradePeriodEnd"), status).second;
```
It might be obvious to others. What does `tpe` stand for?

ripcurlx commented on desktop/src/main/java/bisq/desktop/main/overlays/windows/DisputeSummaryWindow.java:
```diff
@@ -970,4 +983,31 @@ private void applyTradeAmountRadioButtonStates() {
+                dispute.setPayoutTxConfirms(status);
+                displayPayoutStatus(status);
+            }));
+*/
```
I don't think this comment should be checked in.

ripcurlx commented on desktop/src/main/java/bisq/desktop/main/shared/ChatView.java:
```diff
@@ -137,10 +137,12 @@
+    public ChatView(SupportManager supportManager, CoinFormatter formatter, String cptyName) {
         this.supportManager = supportManager;
         this.formatter = formatter;
+        this.cptyName = cptyName;
```
`cpty` stands for?

ripcurlx commented on core/src/main/java/bisq/core/support/dispute/Dispute.java:
```diff
@@ -107,10 +132,25 @@
+    private transient long payoutTxConfirms = -1;
+
+    private transient final BooleanProperty isClosedProperty = new SimpleBooleanProperty();
+    private transient final IntegerProperty alertCountProperty = new SimpleIntegerProperty();
```
🤔 not sure if alert count is the perfect naming. Maybe notificationCount or badgeCount would be more common for this?

ripcurlx reviewed (changes requested):
NACK - Some naming suggestions and comments I would leave out of this PR.

Also I tested the functionality and found following UX issue:

When a new mediation case is opened it looks like that to the user:
![Bildschirmfoto 2021-03-15 um 11 39 47](https://user-images.githubusercontent.com/170962/111140918-2ef5c200-8583-11eb-82e4-f23e03c80245.png)

It is not immediately clear to me what the action I need to do to get rid of the `New` badge (I need to open the chat). For this initial system message I think the chat icon should have a (1) badge already, so users are clicking on it. I also think it would be great to open the chat window on double click of the dispute item as the default action. WDYT? I'm continuing to test the remaining functionality and will add my comments below. But besides that - great improvements already 👍

ripcurlx commented:
![Bildschirmfoto 2021-03-15 um 11 45 49](https://user-images.githubusercontent.com/170962/111141595-04f0cf80-8584-11eb-86c9-fd29722070e8.png)

I think if multiple chat messages are sent from one dispute which can be resolved by looking at one chat window the badge count on support support should be (1).

ripcurlx commented:
I'm also not sure if the cog icon is the perfect choice for the process. I always think about settings or preferences when I see it. Maybe `CLIPBOARD_CHECK`
![Bildschirmfoto 2021-03-15 um 11 55 17](https://user-images.githubusercontent.com/170962/111142796-5ea5c980-8585-11eb-8045-bb41dd197682.png)
or `FORMAT_LIST_CHECKS` could be a better choice. WDYT?
![Bildschirmfoto 2021-03-15 um 11 53 54](https://user-images.githubusercontent.com/170962/111142801-5fd6f680-8585-11eb-9494-a2155a636fd2.png)

Commit:
code review changes requested by @ripcurlx

renamed alertCountProperty to badgeCountProperty
changed the badge counting policy so that the count of chat messages does not contribute to the badge number, but merely the presence of chat messages, and/or the new flag.
show an explanatory message when trader is not allowed to close own dispute ticket
removed some comments and code that were accidentally included from PR5160
renamed variable tpe to tradePeriodEnd
renamed variable cptyName to counterpartyName
clear the new badge on double-click of dispute tablerow

ghost commented:
@ripcurlx Thanks for the detailed review!  I've addressed all issues and have an open question with @leo816 and @huey735 about the Process icon. 

https://github.com/bisq-network/bisq/pull/5207/commits/ad4927a666c9c5fa65d298499e338bbcd87754b7

- renamed alertCountProperty to badgeCountProperty
- changed the badge counting policy so that the count of chat messages does not contribute to the badge number, but merely the presence of chat messages, and/or the new flag.
- show an explanatory message when trader is not allowed to close own dispute ticket
- removed some comments and code that were accidentally included from PR5160
- renamed variable tpe to tradePeriodEnd
- renamed variable cptyName to counterpartyName
- clear the new badge on double-click of dispute tablerow

Commit:
double-click opens chat window

ghost commented:
:+1:   @ripcurlx I implemented latest requested changes.
No response from @huey735 or @leo816 about the process icon, & I prefer the cog.

leo816 commented:
that seems great, sorry I didn't answer earlier, having a lot of conversations at once and missed this. Good Job.

huey735 commented:
That's amazing work @jmacxx 
Regarding the `Process` icon, I share @ripcurlx 's sentiment. The cog makes me think of `Settings`. Maybe a gavel makes more semantic sense.

![gavel002](https://user-images.githubusercontent.com/44140017/111390428-a07c6000-86aa-11eb-9453-d3a924ee0f09.png)

Or if that's strong, maybe a balance?

![balance](https://user-images.githubusercontent.com/44140017/111390317-72971b80-86aa-11eb-8493-bcac63cb6939.png)

@ripcurlx suggestions don't seem thematically relevant. But then again I imagine that with time the chosen icon will gain meaning.

Commit:
process icon use GAVEL

ghost commented:
![image](https://user-images.githubusercontent.com/47253594/111397408-37d3ba00-868f-11eb-9ac9-775dc8b34c94.png)

pazza83 commented:
great work @jmacxx many thanks

ripcurlx commented:
Great work @jmacxx :+1:
